### ISSUE TITLE
httpResponse.earlyHints includes unneeded JS in HTML-only mode

### ISSUE THREAD

**Issue by jonaskuske** [vikejs/vike#1574]
### Description

I'm using HTML-only mode, which loads no JS besides the `+client.ts` files. Still, earlyHints instruct the browser to preload not only the `+clients.ts` script, but also the regular `index.ts` and corresponding chunks.

This leads to overfetching and logs warnings in the console:

![image](https://github.com/vikejs/vike/assets/30421456/c0e26869-93d2-4b97-bc21-76c0bfda138f)

**Comment by jonaskuske** [comment]
Workaround:

```ts
const hintSources = new Set(httpResponse.earlyHints.map((hint) => hint.src))
const pageEntry = Object.values(assetsManifest!).find(
  (entry) => entry.src?.includes('pageConfigValuesAll') && hintSources.has('/' + entry.file),
)
const pageEntrySources = new Set(
  [pageEntry?.file].concat(pageEntry?.imports ?? []).map((file) => '/' + file),
)

const filteredEarlyHints = httpResponse.earlyHints.filter(hint => !pageEntrySources.has(hint.src))

await reply.writeEarlyHintsLinks?.(
  filteredEarlyHints.map(hint => hint.earlyHintLink)
)
```

**Comment by brillout** [comment]
Indeed, I can reproduce. How do you get the warnings in the bowser console? I don't get any.

**Comment by jonaskuske** [comment]
Hmm, I don't know, they just show up after a couple of seconds for me in Edge, should work the same in Chrome though. Have you checked whether EH work at all?
(in console: `performance.getEntries().filter(e => e.initiatorType === 'early-hints')`)

**Comment by brillout** [comment]
Actually, the two early hints I can reproduce are CSS early hints which is correct.

I cannot reproduce JavaScript early hints, see https://github.com/vikejs/vike/commit/b1d6b4844cef86454f430b2af5666ff6c97406ea.

Can you provide a reproduction?

**Comment by brillout** [comment]
Ah, wait, actually I can reproduce for a production build. Let me fix this.

**Comment by jonaskuske** [comment]
Looking at the assertion, is your plan to exclude all JS in HTML-only mode? `+client.ts` and its imports can stay, no?

**Comment by brillout** [comment]
Yes, early hints for `client.js` should stay.

(FYI I get an empty array upon `performance.getEntries().filter(e => e.initiatorType === 'early-hints')`, I guess my Express.js or Chrome setup doesn't support early hints :thinking:)

**Comment by brillout** [comment]
Fix pre-released as `0.4.167-commit-14e4b5e`.

Btw. in case your company is up for it [Vike is looking for sponsors](https://github.com/vikejs/vike/issues/1350) &mdash; it makes a massive difference.

**Comment by jonaskuske** [comment]
Thank you! And would love to – but unfortunately, I only use Vike privately (for my thesis on app render modes/architecture https://thesis.joku.co) and don't have the means to sponsor :(
